File: Programmers/codingTest_Kit/Hash/42579.cpp
#include <string>
#include <vector>
#include <unordered_map>
#include <iostream>
#include <algorithm>
using namespace std;
bool cmp(pair<int, int> a, pair<int, int> b){
    return a.first > b.first;
}
vector<int> solution(vector<string> genres, vector<int> plays) {
    vector<int> answer;
    unordered_map<string, unordered_map<int, int>> musicList;
    unordered_map<string, int> music;
    int N = genres.size();
    for(int i = 0 ; i < N; i++) {
        musicList[genres[i]][i] = plays[i];
        music[genres[i]] += plays[i];
    }
    while(music.size()){
        string genre;
        int max = 0;
        for(auto mu : music){
            if(max < mu.second){
                max = mu.second;
                genre = mu.first;
            }
        }
        for(int i = 0; i < 2; i++){
            int max = 0, index = -1;
            for(auto data : musicList[genre]){
                if(max < data.second){
                    max = data.second;
                    index = data.first;
                }
            }
            if(index != -1){
                answer.push_back(index);
                musicList[genre].erase(index);
            }
            else break;
        }
        music.erase(genre);
    }
    return answer;
}

int main(){
    solution({"classic", "pop", "classic", "classic", "pop"},{500, 600, 150, 800, 2500});
}

File: Programmers/codingTest_Kit/Hash/42579.py
def solution(genres, plays):
    answer = []
    hs = dict()
    for i in range(len(genres)):
        if genres[i] in hs:
            hs[genres[i]] += plays[i]
        else:
            hs[genres[i]] = plays[i]

    while len(hs) > 0:
        genre_max = max(hs.keys(), key=(lambda k : hs[k]))
        del(hs[genre_max])

        second = largest = 0
        for i in range(len(genres)):
            if genres[i] == genre_max:
                if plays[i] >= largest:
                    second = largest
                    largest = plays[i]
                elif second < plays[i] < largest:
                    second = plays[i]
        answer.append(plays.index(largest))
        if second != 0:
            answer.append(plays.index(second))
    return answer